pkg/reconciler/managed/connector_adapter.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package managed
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+type missingRef struct{ error }
+
+func (m missingRef) MissingReference() bool { return true }
+
+// IsMissingReference returns true if an error indicates that a managed
+// resource is missing a required reference..
+func IsMissingReference(err error) bool {
+	_, ok := err.(interface {
+		MissingReference() bool
+	})
+	return ok
+}
+
+// A ProviderConfigConnector connects to a provider using information extracted
+// from the supplied provider configuration and managed resource.
+type ProviderConfigConnector interface {
```

Comment:
I think that assumes that the `ProviderConfigAdapter` or whatever middleware we use would be responsible for creating the PCU, but not for getting and passing the PC. That feels weird to me - I'd expect us to either be 'full service' and handle both, or handle neither. I suspect the idea that Crossplane should automate the management of PCUs but not automate the fetching of PCs might be would be unusual to folks who aren't as used to the way it works today as we are.

Reply:
> I suspect the idea that Crossplane should automate the management of PCUs but not automate the fetching of PCs might be would be unusual to folks who aren't as used to the way it works today as we are.

Conceptually, I see two operations (1) PCU machinery (2) connecting to cloud provider. Fetching the `PC` happens to be a common operation for them but I can't really see it as something weird for both of them doing it separately if that makes the code simpler by not introducing a new interface that is very similar to an existing one. If you feel strongly that `PC` fetching will always happen all the time when managed reconciler is used, we can discuss changing `ExternalConnector` interface to how `ProviderConfigAdapter` looks, i.e. `Connect(ctx context.Context, pc resource.Provider, mg resource.Managed) (ExternalClient, error)`. Though anecdotally, that would've made it more painful to do `Provider` -> `ProviderConfig` if we had designed it that way in the beginning.

In fact, I could even be convinced that managed reconciler could have a separate call than `Connect` to do the PCU operation if we'd like to have it abstracted with its own interface. One is completely Crossplane specific and the other one is completely cloud-related.

Reply:
I honestly feel there's a bias here that handling a PCU is "machinery" that should be done for the provider author while the act of finding and fetching a PC is not. I think the only difference here is that PCU is new and controversial, while finding and fetching a PC is an established pattern that is second nature to us.

That said, I've abandoned the approach of building an adaptor that automates PCU tracking and PC fetching. It could be worth pursuing that approach in future, but it felt like too much abstraction while we still need to deal with the deprecated Provider approach.